cpp/src/arrow/compute/kernels/filter.h

```diff
@@ -57,6 +79,13 @@ class ARROW_EXPORT FilterBinaryKernel : public BinaryKernel {
   Status Call(FunctionContext* ctx, const Datum& left, const Datum& right,
               Datum* out) override;
 
+  static int64_t out_length(const Datum& left, const Datum& right) {
+    if (left.kind() == Datum::ARRAY) return left.length();
+    if (right.kind() == Datum::ARRAY) return right.length();
+
+    return 0;
```

Comment:
does it pay to support scalar-scalar comparison (i.e. should this be 1)?

Reply:
I think we can support comparing scalars by promoting Scalar to Array of length 1. Is there a JIRA about Scalar->Array promotion?

Reply:
https://jira.apache.org/jira/browse/ARROW-4567